File: server/src/game/cord.rs
use std::iter::FromIterator;
use std::collections::VecDeque;
use std::cmp::max;
use std::ops;

#[derive(Serialize, Deserialize, Copy, Clone, PartialEq, Debug)]
pub struct Cord(pub isize, pub isize, pub isize);

impl Cord {
    const zero: Cord = Cord(0,0,0);

    pub fn is_linear_to(self, other: Cord) -> bool {
        let mut v = [(self.0-other.0).abs(),
            (self.1-other.1).abs(),
            (self.2-other.2).abs()];
        v.sort();
        v[0] == 0 && v[1] == v[2]
    }

    pub fn distance(self, other: Cord) -> usize {
        max(max((self.0-other.0).abs(), (self.1-other.1).abs()), (self.2-other.2).abs()) as usize
    }

    pub fn dir(self, other: Cord) -> Cord {
        let n = self.distance(other) as isize;
        let t = other - self;
        t / n
    }

    pub fn vec_size(self) -> usize {
        Cord::zero.distance(self)
    }

    pub fn is_linear_vec(self) -> bool {
        Cord::zero.is_linear_to(self)
    }

    pub fn linedraw(self, other: Cord) -> Vec<Cord> {
        let n = self.distance(other);
        let mut out: VecDeque<Cord> = VecDeque::new();
        for i in 0..n+1 {
            out.push_back(self.cube_lerp(other, 1.0 / n as f32 * i as f32));
        }
        Vec::from_iter(out.into_iter())
    }

    fn lerp(a: isize, b: isize, t: f32) -> isize {
        a + ((b - a) as f32 * t) as isize
    }

    fn cube_lerp(self, other: Cord, t: f32) -> Cord {
        Cord(Cord::lerp(self.0, other.0, t), 
                Cord::lerp(self.1, other.1, t),
                Cord::lerp(self.2, other.2, t))
    }

}

impl ops::Add<Cord> for Cord{
    type Output = Cord;

    fn add(self, _rhs: Cord) -> Cord {
        Cord(self.0 + _rhs.0, self.1 + _rhs.1, self.2 + _rhs.2)
    }
}

impl ops::Sub<Cord> for Cord{
    type Output = Cord;

    fn sub(self, _rhs: Cord) -> Cord {
        Cord(self.0 - _rhs.0, self.1 - _rhs.1, self.2 - _rhs.2)
    }
}

impl ops::Mul<isize> for Cord{
    type Output = Cord;

    fn mul(self, _rhs: isize) -> Cord {
        Cord(self.0 * _rhs, self.1 * _rhs, self.2 * _rhs)
    }
}

impl ops::Div<isize> for Cord{
    type Output = Cord;

    fn div(self, _rhs: isize) -> Cord {
        Cord(self.0 / _rhs, self.1 / _rhs, self.2 / _rhs)
    }
}

impl ops::Neg for Cord {
    type Output = Cord;

    fn neg(self) -> Cord {
        self * -1
    }
}

File: ai/ai/game/cord.py
from typing import List

class Cord:
    def __init__(self, x: int, y: int, z: int):
        self.x = x
        self.y = y
        self.z = z

    def __add__(self, other: 'Cord') -> 'Cord':
        return Cord(self.x + other.x, self.y + other.y, self.z + other.z)

    def __neg__(self) -> 'Cord':
        return Cord(-self.x, -self.y, -self.z)

    def __sub__(self, other: 'Cord') -> 'Cord':
        return self + (-other)

    def __floordiv__(self, t: int) -> 'Cord':
        return Cord(self.x // t, self.y // t, self.z // t)

    def is_linear_to(self, other: 'Cord') -> bool:
        v = [abs(self.x - other.x), abs(self.y - other.y), abs(self.z - other.z)]
        v = sorted(v)
        return v[0] == 0 and v[1] == v[2]
    
    def distance(self, other: 'Cord') -> int:
        return max(abs(self.x-other.x), abs(self.y-other.y), abs(self.z-other.z))
    
    def dir(self, other: 'Cord') -> 'Cord':
        n = self.distance(other)
        t = other - self
        return t // n

    def vec_size(self) -> int:
        return Cord(0,0,0).distance(self)

    def is_linear_vec(self) -> bool:
        return Cord(0,0,0).is_linear_to(self)

    def linedraw(self, other: 'Cord') -> List['Cord']:
        n = self.distance(other)
        return [self.cube_lerp(other, 1.0 / n * i) for i in range(n+1)]

    @staticmethod
    def lerp(a: int, b: int, t: float) -> int: 
        return int(a + ((b - a) * t))

    def cube_lerp(self, other: 'Cord', t: float) -> 'Cord':
        return Cord(Cord.lerp(self.x, other.x, t), 
                Cord.lerp(self.y, other.y, t),
                Cord.lerp(self.z, other.z, t))